### PR TITLE
Make all CF types Equatable and Hashable.

### PR DESCRIPTION
**Pull request by jrose-apple**
- **Explanation:** Now that we've removed implicit bridging, making a CFDictionary to pass to, say, CoreGraphics is rather difficult to do easily. Ideally we'd want a CFString-keyed dictionary (since all the global constants are CFStrings), but we can't do that without CFString being Hashable. However, _all_ CF types are hashable in the same way as NSObject: there's a standard `CFHash` function that works across types. This commit adds a new protocol, _CFObject, which includes Hashable and Equatable implementations; it also tweaks protocol checking to allow operator requirements to be found in overlays for a particular module. Note that this will break anyone who implemented Hashable for a CF type themselves.
- **Scope:** Affects all imported CF types. Sema change should not affect any existing behavior because no imported types were using protocols with operator requirements that wouldn't be found.
- **Issue:** [SR-2388](https://bugs.swift.org/browse/SR-2388)
- **Reviewed by:** @jckarter and @DougGregor
- **Risk:** Medium-low. Most likely very few people were doing this, and the Sema change _shouldn't_ affect anything.
- **Testing:** Added new regression tests.

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

A smoke test on macOS does the following:
1. Builds the compiler incrementally.
2. Builds the standard library only for macOS. Simulator standard libraries and
   device standard libraries are not built.
3. lldb is not built.
4. The test and validation-test targets are run only for macOS. The optimized
   version of these tests are not run.

A smoke test on Linux does the following:
1. Builds the compiler incrementally.
2. Builds the standard library incrementally.
3. lldb is built incrementally.
4. The swift test and validation-test targets are run. The optimized version of these
   tests are not run.
5. lldb is tested.

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by jrose-apple** [comment]
@swift-ci Please test macOS

**Comment by jrose-apple** [comment]
@parkera, any objections?

File: stdlib/public/SDK/CoreFoundation/CoreFoundation.swift [hunk 1]
```diff
@@ -0,0 +1,23 @@
+//===----------------------------------------------------------------------===//
```
Comment (parkera) [inline stdlib/public/SDK/CoreFoundation/CoreFoundation.swift]: Please name this file something more descriptive than CoreFoundation.swift. CFObject.swift maybe.

Comment (jrose-apple) [inline stdlib/public/SDK/CoreFoundation/CoreFoundation.swift]: Unfortunately our CMake setup currently takes the name of the result file from the name of the first file in the module.

Comment (parkera) [inline stdlib/public/SDK/CoreFoundation/CoreFoundation.swift]: I would then prefer to see an empty CoreFoundation.swift and this stuff in a more descriptive place.

I really don't like how Foundation.swift became a dumping ground for whatever needed to be put there.

Comment (jrose-apple) [inline stdlib/public/SDK/CoreFoundation/CoreFoundation.swift]: Okay, will fix in master. I'll see if the CMake stuff is easy to change, too.

**Comment by gparker42** [comment]
Reverting. New test Interpreter/SDK/cf.swift fails in two places when built with --test-optimized. I don't know if it's a bug here or a miscompile of this code.
https://ci.swift.org/view/swift-3.0-branch/job/oss-swift-3.0_tools-RA_stdlib-RDA_test-simulator/77/

**Comment by gparker42** [comment]
Also reverting #4394 on the 3.0 branch.

**Pull request merged**